### PR TITLE
Working with lambda expressions in Kotlin

### LINKED ISSUE #1603: Working with lambda functions in Kotlin
### Proposed title of article
Working with lambda functions in Kotlin

### Introduction paragraph(s):
In Kotlin, functions are ‘first-class citizens’. This means that you can do the following to functions in Kotlin: 
1.	Functions can be passed as arguments to other functions and classes.
2.	Functions can be assigned to variables and stored in data structures.
3.	Functions can be returned from other higher order functions.
The above functionality is enabled through the use of lambda functions.

Lambda functions are called when the higher order function, (the function that calls the lambda function), is created. This means that by the time we implement the higher order function, the lambda function will already possess the arguments passed to it during creation, and thus we can manipulate those values for our own purposes.

The above paragraph will be demonstrated.

### Key takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article? Use this area to get your ideas down on the bulk of your article or tutorial.
1. What lambda functions are.
2. Why lambda functions are important.
3. When to use lambda functions.

### References: 
https://kotlinlang.org/docs/lambdas.html
https://en.wikipedia.org/wiki/First-class_function

The article will be different from the references in that those references do not explain explicitly why lambda functions are important and they also give vague examples of when to use lambda functions. 
This article will tackle those issues in detail. Lambda functions happen to be a very confusing area to beginners and intermediate developers

### PR DESCRIPTION
**Pull request by RaphaelNdonga**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you placed your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #1603

### CODE REVIEW

**Comment by LinusMuema** [comment]
Hi @RaphaelNdonga 
I will be reviewing your article keep an eye for the updates in this pull request.

File: articles/working-with-lambda-expressions-and-anonymous-functions/index.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
+### Introduction
+
+Lambda expressions can be challenging to understand. Types that look like this: `(Int) -> String` are not intuitive to many developers. This article aims to simplify the concept behind lambda expressions and how to use thm.
+
+### Prerequisites
+
+1. A basic understanding of the Kotlin programming language.
+2. Either IntelliJ IDEA or access to [Kotlin Playground](https://play.kotlinlang.org/)
```
Comment (LinusMuema) [inline articles/working-with-lambda-expressions-and-anonymous-functions/index.md:8]: I noticed you were adding links directly, in the other articles make sure you use the markdown styling for links.

File: articles/working-with-lambda-expressions-and-anonymous-functions/index.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+**Step 3. Calling the function:** here you call the function and give it the parameters it requires to perform its task.
+
+> **Note**: This article refers to functions in the context of functional programming. In object-oriented programming, you can use interfaces and abstract classes. They can postpone function implementation. Later, you can override the function when you implement the interface or inherit the class.
+
+### A use case of lambda expressions
+
+Regular functions are created in the same place where they implement their functionality. Have a look at this function:
+
+```kotlin
+fun add(a:Int, b:Int){
```
Comment (LinusMuema) [inline articles/working-with-lambda-expressions-and-anonymous-functions/index.md:63]: Give your functions a short name to make them more readable and the code to be cleaner. A technical suggestion here.

**Review by LinusMuema — APPROVED** [review]
Hi @RaphaelNdonga 
This is an awesome first article. :+1: 
I noticed two issues though Or rather, the same issue on two items. Make sure you name your images following the section convention. For the author, your profile should be `avatar` and the hero image for the article should be called `hero`.

Otherwise, this article is awesome and ready to be published. :sunglasses:

**Comment by LinusMuema** [comment]
Hi @hectorkambow 
You could give this article a final review. It is ready to go :rocket:

**Comment by hectorkambow** [comment]
great work gentlemen @RaphaelNdonga 

cc @LinusMuema

**Pull request merged**